packages/compartment-mapper/src/assemble.js

```diff
@@ -2,15 +2,110 @@
 /// <reference types="ses" />
 
 import { resolve } from './node-module-specifier.js';
-import { mapParsers } from './parse.js';
+import { parseExtension } from './extension.js';
 
-const { entries, freeze } = Object;
+const { entries, fromEntries, freeze } = Object;
 
 const defaultCompartment = Compartment;
 
 // q, as in quote, for strings in error messages.
 const q = JSON.stringify;
 
+/**
+ * @template {string | number} Key
+ * @template Value
+ * @param {Record<Key, Value>} object
+ * @param {Key} key
+ * @returns {boolean}
+ */
+const has = (object, key) => Object.prototype.hasOwnProperty.call(object, key);
+
+/**
+ * @param {Record<string, string>} languageForExtension - maps a file extension
+ * to the corresponding language.
+ * @param {Record<string, string>} languageForModuleSpecifier - In a rare case,
+ * the type of a module is implied by package.json and should not be inferred
+ * from its extension.
+ * @param {Record<string, ParseFn>} parserForLanguage
+ * @param {ModuleTransforms} transforms
+ * @returns {ParseFn}
+ */
+export const makeExtensionParser = (
+  languageForExtension,
+  languageForModuleSpecifier,
+  parserForLanguage,
+  transforms,
+) => {
+  return async (bytes, specifier, location, packageLocation) => {
+    let language;
+    if (
+      Object(languageForModuleSpecifier) === languageForModuleSpecifier &&
+      has(languageForModuleSpecifier, specifier)
+    ) {
+      language = languageForModuleSpecifier[specifier];
+    } else {
+      const extension = parseExtension(location);
+      if (!has(languageForExtension, extension)) {
+        throw new Error(
+          `Cannot parse module ${specifier} at ${location}, no parser configured for extension ${extension}`,
+        );
+      }
+      language = languageForExtension[extension];
+    }
+
+    if (has(transforms, language)) {
+      ({ bytes, parser: language } = await transforms[language](
+        bytes,
+        specifier,
+        location,
+        packageLocation,
+      ));
+    }
+
+    if (!has(parserForLanguage, language)) {
+      throw new Error(
+        `Cannot parse module ${specifier} at ${location}, no parser configured for the language ${language}`,
+      );
+    }
+    const parse = parserForLanguage[language];
+    return parse(bytes, specifier, location, packageLocation);
+  };
+};
+
+/**
+ * @param {Record<string, ParserDescriptor>} parsers
+ * @param {Record<string, string>} languageForModuleSpecifier - In a rare case, the type of a module
+ * is implied by package.json and should not be inferred from its extension.
+ * @param {Record<string, ParseFn>} parserForLanguage
+ * @param {ModuleTransforms} transforms
+ * @returns {ParseFn}
+ */
+export const mapParsers = (
+  parsers,
```

Comment:
```suggestion
  parserDescs,
```
maybe? Since you also have genuine parsers, this confused me.